dj/api/nodes.py

```diff
@@ -113,17 +146,208 @@ def validate_node(
+            for column_name, column_data in data.columns.items()
+        ],
+        tables=[
+            Table(
+                catalog=table.catalog,
+                schema_=table.schema_,
+                table=table.table,
+                cost=table.cost,
+                database=databases[table.database_name],
+                columns=[],
+            )
+            for table in data.tables
+        ],
+        parents=[],
+    )
+    return node_revision
+
+
+def create_node_revision(data: CreateNode, session: Session) -> Node:
     """
-    List the available nodes.
+    Create a non-source node revision.
     """
-    return session.exec(select(Node)).all()
+
+    node_revision = Node.parse_obj(data)
+    new_parents = (
+        get_dependencies(node_revision.query) if node_revision.query else set()
+    )
+    parent_refs = session.exec(
+        select(ReferenceNode).where(
+            # pylint: disable=no-member
+            ReferenceNode.name.in_(  # type: ignore
+                new_parents,
+            ),
+        ),
+    ).all()
+    node_revision.parents = [ref.current for ref in parent_refs]
+
+    _logger.info(
+        "Parent nodes for %s (v%s): %s",
+        data.name,
+        node_revision.version,
+        node_revision.parents,
+    )
+
+    # Infer the new node revision's column names and column types by inspecting the parent nodes
+    try:
+        node_revision.columns = infer_columns(
+            node_revision.query,
+            parents=node_revision.parents,
+        )
+    except Exception as exception:  # pylint: disable=broad-except
+        error_message = f"Cannot infer columns for {data.name}: {str(exception)}"
+        _logger.error(error_message)
+    return node_revision
+
+
+@router.post("/nodes/", response_model=RefNodeMetadata)
+def create_node(
+    data: Union[CreateSourceNode, CreateNode],
+    session: Session = Depends(get_session),
+) -> RefNodeMetadata:
+    """
+    Create a node.
+    """
+    query = select(ReferenceNode).where(ReferenceNode.name == data.name)
+    ref_node = session.exec(query).one_or_none()
+    if ref_node:
+        raise DJException(f"A node with name `{data.name}` already exists.")
+
+    ref_node = ReferenceNode(name=data.name, type=data.type, current_version=0)
+
+    if data.type == NodeType.SOURCE:
+        node_revision = create_source_node_revision(data, session)
+    else:
+        node_revision = create_node_revision(data, session)
+
+    # Point the reference node to the new node revision.
+    node_revision.reference_node = ref_node
+    node_revision.version = ref_node.current_version + 1
+    ref_node.current_version = node_revision.version
+
+    node_revision.extra_validation()
+
+    session.add(ref_node)
+    session.commit()
+    session.refresh(ref_node)
+    return ref_node  # type: ignore
+
+
+@router.put("/nodes/{name}/", response_model=RefNodeMetadata)
+def update_node(
+    name: str,
+    data: Union[CreateSourceNode, NodeBase],
```

Comment:
Since `NodeBase` requires `name: str`, it requires you to include the name twice when making an API call (the other one being the `{name}` in the route). We could remove the route parameter but if we want to keep that maybe we need to create an `UpdateNode` class that's basically `NodeBase` without the name attribute.

Reply:
Yep, I added an `UpdateNode` class, and since it's being used with `patch` instead of `put`, it actually just modifies the mutable fields to be `Optional` so that only a subset of them need to be provided.